hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieWriteHandle.java

```diff
@@ -103,9 +114,34 @@ protected HoodieWriteHandle(HoodieWriteConfig config, String instantTime, String
     this.taskContextSupplier = taskContextSupplier;
     this.writeToken = makeWriteToken();
     this.schemaOnReadEnabled = !isNullOrEmpty(hoodieTable.getConfig().getInternalSchema());
+    this.preserveMetadata = preserveMetadata;
     this.recordMerger = config.getRecordMerger();
     this.writeStatus = (WriteStatus) ReflectionUtils.loadClass(config.getWriteStatusClassName(),
         hoodieTable.shouldTrackSuccessRecords(), config.getWriteStatusFailureFraction(), hoodieTable.isMetadataTable());
+    boolean isMetadataStreamingWritesEnabled = config.isMetadataStreamingWritesEnabled(hoodieTable.getMetaClient().getTableConfig().getTableVersion());
+    if (isMetadataStreamingWritesEnabled) {
+      initMetadataPartitionsToCollectStats(preserveMetadata);
+      this.isSecondaryIndexStatsStreamingWritesEnabled = !secondaryIndexDefns.isEmpty();
+    } else {
+      this.isSecondaryIndexStatsStreamingWritesEnabled = false;
+    }
+  }
+
+  private void initMetadataPartitionsToCollectStats(boolean preserveMetadata) {
+    // Secondary index should not be updated for clustering and compaction
+    // Since for clustering and compaction preserveMetadata is true, we are checking for it before enabling secondary index update
+    if (config.isSecondaryIndexEnabled() && !preserveMetadata) {
+      ValidationUtils.checkArgument(recordMerger.getRecordType() == HoodieRecord.HoodieRecordType.AVRO,
+          "Only Avro record type is supported for streaming writes to metadata table with write handles");
+      secondaryIndexDefns = hoodieTable.getMetaClient().getIndexMetadata()
```

Comment:
instead of maintain the pairs of `(indexName -> index definition)`, we should maintain the `(indexName -> index source fields)` by using the

`String.join(".", secondaryIndexPartitionPathFieldPair.getValue().getSourceFields())`

to avoid unnecessary computation per-record.